numpy/array_api/linalg.py

```diff
@@ -391,18 +392,43 @@ def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, int]]] = No
     if x.dtype not in _floating_dtypes:
         raise TypeError('Only floating-point dtypes are allowed in norm')
 
+    # np.linalg.norm tries to do a matrix norm
     a = x._array
     if axis is None:
-        a = a.flatten()
-        axis = 0
+        # Note: np.linalg.norm() doesn't handle dimension 0 arrays
+        if a.ndim == 0:
+            a = a[None]
+        else:
+            a = a.flatten()
+        _axis = 0
     elif isinstance(axis, tuple):
-        # Note: The axis argument supports any number of axes, whereas norm()
-        # only supports a single axis for vector norm.
-        rest = tuple(i for i in range(a.ndim) if i not in axis)
+        # Note: The axis argument supports any number of axes, whereas
+        # np.linalg.norm() only supports a single axis for vector norm.
+        normalized_axis = [i if i >= 0 else i + a.ndim for i in axis]
+        rest = tuple(i for i in range(a.ndim) if i not in normalized_axis)
         newshape = axis + rest
-        a = np.transpose(a, newshape).reshape((np.prod([a.shape[i] for i in axis]), *[a.shape[i] for i in rest]))
-        axis = 0
-    return Array._new(np.linalg.norm(a, axis=axis, keepdims=keepdims, ord=ord))
-
+        a = np.transpose(a, newshape).reshape(
+            (np.prod([a.shape[i] for i in axis], dtype=int), *[a.shape[i] for i in rest]))
+        _axis = 0
+    else:
+        _axis = axis
+
+    res = Array._new(np.linalg.norm(a, axis=_axis, ord=ord))
+
+    if keepdims:
+        # We can't reuse np.linalg.norm(keepdims) because of the hacks above
+        # to avoid matrix norm logic.
+        shape = list(x.shape)
+        if axis is None:
+            _axis = range(x.ndim)
+        elif isinstance(axis, int):
+            _axis = (axis,)
+        else:
+            _axis = axis
```

Comment:
Might put this together with the setup (i.e. figuring out the result shape there?  But looks fine.